src/transformers/generation/utils.py

```diff
@@ -4819,14 +4819,23 @@ def _speculative_sampling(
     p_n_plus_1 = p[:, n_matches, :]
     if n_matches < gamma:
         q_n_plus_1 = q[:, n_matches, :]
-        p_prime = torch.clamp((p_n_plus_1 - q_n_plus_1), min=0).softmax(dim=-1)
+        p_prime = torch.clamp((p_n_plus_1 - q_n_plus_1), min=0)
```

Comment:
Context for @amyeroberts: the paper normalizes this vector, as opposed to a softmax 🙃

Reply:
I don't think it is necessary, if `p_n_plus_1 == q_n_plus_1` then you would have accepted this token in the first place because `p_i / q_i == 1` in this case. From what I see the problem is caused when we generate more candidate tokens then `max_length`.

Reply:
Ah, I see, you escape when max_new_tokens is 0.

Let me see if I can find a prettier way of encoding that clause :)

Reply:
@ofirzaf I see what's happening. In fact, we need several fixes 🤗 Your solution works in most cases. However, there would still be one corner case when the candidate generator generates a sequence ending with EOS token AND the main model accepts that token. In a correct setup, generation would stop there (we shouldn't generate beyond an EOS token), but the current algorithm is asking the main model to resample the last token (instead of selecting the EOS token) in that situation.

I believe the following will work:
1. Update the candidate model's `max_new_tokens` as you described. Remove your other changes in the last two commits;
2. Update `gamma` to `gamma = min(candidate_logits.shape[1], max_matches)`. This ensures we don't accept more candidates than we should, as `max_matches` contains the maximum admissible length discounting the token that will be sampled from the main model;
3. Inside `if last_assistant_token_is_eos and n_matches == candidate_length:`, add `valid_tokens = new_candidate_input_ids[:, :n_matches + 1]` after `n_matches -= 1`. In a nutshell, don't resample from the main model when the last accepted token is an EOS while keeping `n_matches` consistent with the logic to correctly slice optional output tensors such as `past_key_values`;
4. All lines between `n_matches = min(n_matches, max_matches)` and `valid_tokens = t` should be inside an `else:` block -- an `else` that opposes to `if last_assistant_token_is_eos and n_matches == candidate_length:` clause. This ensures 3. works correctly and jumps straight to the `return` when triggered.

Since you've got a keen eye for speculative decoding, I'd like to kindly ask to review these proposed changes 🤗

Reply:
Sounds good. Do you still need to do `n_matches -= 1`? Since we do accept the match of the EOS.
https://github.com/huggingface/transformers/blob/190a63a0b9741cf7c2596557f11eb4da6203c04f/src/transformers/generation/utils.py#L4815
<p>
Another thing that I tried to fix which now happens with this solution is the following case:
Suppose in `get_candidates` you get `max_length - new_cur_len - 1 == 0`, in this case you still generate a single candidate (I guess `generate` ignores `max_new_tokens`) and then the target model will get this token as an input, creating new logits for 2 tokens. Then, speculative will reject/accept this token but since `max_matches == 0` we will ignore that and sample a token from the target.<br>
So now the solution is correct but in this corner case you do spend unnecessary compute for logits/tokens you won't use by definition.<br>
In the last commit that I deleted I avoided that by skipping the generation of candidate tokens in case there is no need.
<p>
What do you think?

Reply:
We do need that line (`n_matches -= 1`), as some output slicing code (e.g. [here](https://github.com/huggingface/transformers/blob/c662c78c71e693234725d70a6291398b33042444/src/transformers/generation/utils.py#L4435)) assumes the output of the model at a given generation step is `n_matches + 1`.

We should add a comment on this line, though, mentioning why it exists!

Reply:
Re corner cases you've identified: tough one. On one hand, we save some computation on the smaller model. On the other hand, it introduces a few extra if/elses, hurting readability. 

Since the point of speculative decoding is performance, let's fix the corner case as you suggested. One extra request, though: on the added if/elses to handle this corner cases, add a small comment explaining why it's there :)

Reply:
Found a way to avoid an extra if :)